feat!: Enable GPU rendering from the submitter

miabatta opened the pull request:
Fixes #148. 

### Edits 02/01/25:
- Changed the new parameter's name to `gpu_device`.
- Exposed GPU rendering & device selection in the submitter UI (auto-populated based on the user & scene settings).

![image](https://github.com/user-attachments/assets/04fa8362-dd79-4418-8b0e-250deb9bcc65)

- Adaptor now iterates through Blender's devices if the supplied GPU device isn't available.

![image](https://github.com/user-attachments/assets/98c1b9dc-1028-44ba-accd-1085744d7582)

- Other misc. code refactoring as suggested!

### What was the problem/requirement? (What/Why)
Nothing in the submitter/adaptor enables GPU rendering based on the user's preferences or scene settings, meaning scenes with GPU rendering enabled will still render using the CPU on SMF.

### What was the solution? (How)
- Add the parameter `GPUDevice` to the default job template. This has a default value of `NONE` (i.e., use the CPU) and accepts any value.
- Edited the adaptor schema to have a `gpu_device` attribute.
- Added a new action type, `set_gpu_device`, to the adaptor.
- Allow users to choose from Blender's natively-supported GPU device types, or manually enter their own. If they have enabled GPU rendering but their chosen device isn't available on the worker, the worker will iterate through Blender's supported device types until it finds one with compatible devices. Fall back to CPU rendering IFF no supported GPU devices are available.

### What is the impact of this change?
GPU rendering is available through the Blender submitter without extra setup!

### How was this change tested?
- Added/updated submitter unit tests to account for the new parameter.
- Manually tested GPU renders with all four device types on both 3.6 and 4.2. Compare the times to render 3 frames:

![image](https://github.com/user-attachments/assets/f1fbd7c7-e914-44b8-a495-e586ce6950eb)
![image](https://github.com/user-attachments/assets/4958e03d-bc6d-4b1f-b061-b4d56a45334a)

### Was this change documented?
Relevant docstrings were updated.

### Is this a breaking change?
**Yes**, this adds a new attribute to the adaptor schema.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

edwards-aws commented on src/deadline/blender_adaptor/BlenderAdaptor/schemas/init_data.schema.json:
```diff
@@ -12,6 +12,15 @@
         },
+        "render_device": {
+            "enum": [
+                "NONE",
```
Since `NONE` in this instance is always going to mean `CPU` what do you think about this being `CPU` instead?

miabatta replied:
This is actually an enum that Blender defines in its user preferences:
![image](https://github.com/user-attachments/assets/dc5e5b97-33b4-4870-a2ac-d81ed5a7bd91)

So there's one specific case where the user has GPU set in the scene, but "None" set in preferences, so `NONE` will be added to the job template [in `template_filling` line 405](https://github.com/aws-deadline/deadline-cloud-for-blender/pull/159/files#diff-14207e3cc2d1f53c6ec708b177bf44c8654104a332eb11b6b8c36aca62514b6cR405). 

I agree it's kind of misleading- I could either change the parameter name to `gpu_device`/`render_gpu_device`/etc, or add some more conditional stuff to make sure `NONE` is always discarded as a value. Latter may work better considering your other comment about the key error; I'll change it when I update the PR :)

edwards-aws replied:
Ahhhhh, I see. Ya I think renaming to something like `gpu_render_device` would go a long way to help with the confusion.

edwards-aws commented on src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py:
```diff
@@ -218,3 +219,35 @@ def set_scene_file(self, data: dict):
+            RuntimeError: If no devices of the specified type are available
+        """
+
+        if not data["render_device"] == "NONE":
```
What do you think about doing an early return here? I generally find less indents easier to read and conditionals without negatives easier to reason about. IMO it also makes it easier to see that when the device is `NONE` nothing is supposed to happen. This is more of a stylistic thing, so it's up to you.

```python
if data["render_device"] == "NONE":
    return

cycles_preferences = bpy.context.preferences.addons["cycles"].preferences
cycles_preferences.refresh_devices()
devices = cycles_preferences.devices

if not devices:
    raise RuntimeError(
        f"Device type {data['render_device']} is not available on this worker."
    )

for device in devices:
    if device.type == "CPU":
        device.use = False
    else:
        device.use = True
        print(f"Enabled GPU device: {device.name}")

cycles_preferences.compute_device_type = data["render_device"]
print(f"Set render device to {data['render_device']}")
bpy.context.scene.cycles.device = "GPU"
```

miabatta replied:
Yeah I wouldn't mind! Thank you for the suggestion

edwards-aws commented on test/unit/deadline_submitter_for_blender/test_submitter.py:
```diff
@@ -368,49 +381,72 @@ def test_get_param_values(submitter_settings, common_layer_settings):
-    )
-    assert filled == expected
+    # Patching out the scene cycles device setting will not add the CyclesRenderDevice parameter
+    with patch("bpy.context.scene.cycles.device", "CPU"):
```
I know that this was already the state of these tests. But I worry over time this will balloon in size. What do you think about either breaking this out into distinct tests or turning this into a parameterized test? This will make it easier to reason about test failures and we'll set ourselves up well for the future.

miabatta replied:
I can definitely do that. As an aside I'm not super happy about the state of these tests in general and might see about refactoring them sometime 🤔

edwards-aws commented on src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py:
```diff
@@ -218,3 +219,35 @@ def set_scene_file(self, data: dict):
+            RuntimeError: If no devices of the specified type are available
+        """
+
+        if not data["render_device"] == "NONE":
```
If someone is using an older submitter this will result in a `KeyError` and crash the adaptor. I'm not sure if that's the behaviour we want here. If the `render_device` key is missing should we print a warning and keep carrying on?

miabatta replied:
That seems reasonable. I'm a little concerned about users expecting GPUs to be automatically enabled in earlier submitter versions, then wasting time rendering on CPUs, for which we _could_ fail the job instead...but at the same time, wasting time rendering on CPUs was the default behaviour beforehand, so a warning may be fine. Curious if others have thoughts!

miabatta committed:
feat!: Enable GPU rendering from submitter

BREAKING CHANGE: This feature adds new functions to the default job
template & adaptor. To use this adaptor version, please make sure you
are using the latest version of the submitter.

Signed-off-by: Mia Battad <133708990+miabatta@users.noreply.github.com>

edwards-aws commented on src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py:
```diff
@@ -218,3 +219,43 @@ def set_scene_file(self, data: dict):
+            _logger.warning(f'Attempting to use "{gpu_device}" instead.')
+            devices = self._list_compatible_render_devices(cycles_preferences, gpu_device)
+
+        bpy.context.scene.cycles.device = "GPU"
```
Should this be `CPU` if the `gpu_device` is `NONE`?

edwards-aws commented on src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py:
```diff
@@ -218,3 +219,43 @@ def set_scene_file(self, data: dict):
+        # https://docs.blender.org/manual/en/4.2/render/cycles/gpu_rendering.html
+        GPU_DEVICE_TYPES_ITER = iter(["OPTIX", "CUDA", "HIP", "ONEAPI", "METAL"])
+
+        gpu_device = data.get("gpu_device", "NONE")
```
Can we do an early exit here if the `gpu_device` is `NONE`? That means we're using a CPU right?

miabatta replied:
Yeah, that was forgetfulness on my part. The check & early return will take care of your other comment re: setting the cycles device to `CPU`.

miabatta commented:
For visibility: I also talked offline with Cody about putting the adaptor code in the cycles handler rather than the default blender handler. I'm expecting that to be the last major change.

evanspearman-a commented on src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/default_blender_template.yaml:
```diff
@@ -65,6 +65,13 @@ parameterDefinitions:
+  userInterface:
+    control: LINE_EDIT
+    label: GPU Device
+  description: Choose the GPU device type to render with when using the cycles engine.
```
Nit: Stylistically I think the word "Choose" should be omitted.

miabatta replied:
Agreed, I don't mind changing it, especially if it stays a line_edit. As an aside, I could also go in and change the other parameters that use this convention another time.

evanspearman-a commented on src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/default_blender_template.yaml:
```diff
@@ -65,6 +65,13 @@ parameterDefinitions:
+- name: GPUDevice
+  type: STRING
+  userInterface:
+    control: LINE_EDIT
```
Any reason we're not using a drop down here?

miabatta replied:
Mostly to be inclusive of other potential setups; I wanted to avoid adding the `ALLOWED_VALUES` attribute if possible, which is required for dropdown parameters. It's a "just-in-case" measure if users have addons/workflows/etc. that would use other GPU device types.

edwards-aws commented on src/deadline/blender_adaptor/BlenderClient/render_handlers/cycles_blender_handler.py:
```diff
@@ -1,7 +1,56 @@
+            _logger.warning(f'Attempting to use "{gpu_device}" instead.')
+            devices = self._list_compatible_render_devices(cycles_preferences, gpu_device)
+
+        bpy.context.scene.cycles.device = "GPU"
```
Do we have an edge case here where the iterator gets to the end and `gpu_device` is `NONE`? I think this should be `CPU` in that case?

miabatta replied:
Good catch, thanks! Should be fast to fix up.

edwards-aws reviewed (approved):
Thanks Mia! This looks good to me!
